### ISSUE TITLE
15/10/2025 - VAE latent space and jacobian calculations

### ISSUE THREAD

**Issue by darin-momayezi** [darin-momayezi/Spark-cells#6]
[spark_vae.ipynb](https://github.com/user-attachments/files/22932040/spark_vae.ipynb)
[spark_vae_jacobian.ipynb](https://github.com/user-attachments/files/22932062/spark_vae_jacobian.ipynb)

This is the first iteration of the spark VAE (trained on reconstruction, classification and KL loss), which takes the spark flow cytometry data and maps it onto a 2D curved manifold. The jacobian calculation of the decoder is also included to calculate the pullback of the decoder, which gives the infinitesimals defined on the tangent space of the curved latent space manifold. Check the papers below for details.

[LATENT SPACE ODDITY: ON THE CURVATURE
OF DEEP GENERATIVE MODELS](https://arxiv.org/pdf/1710.11379)

[A Geometric Perspective on Variational Autoencoders](https://proceedings.neurips.cc/paper_files/paper/2022/file/7bf1dc45f850b8ae1b5a1dd4f475f8b6-Paper-Conference.pdf)

Additions:

A VAE to calculate a meaningful latent space for the flow cytometry data.
A jacobian calculator to visualize the curved manifold on which the latent data lives.
Current issues:

The model is underfitting for the reconstruction task. Notice that the validation loss doesn't change much (or at all).
The manifold is not very smooth. Perhaps increasing the weight of the KL divergence would give a smoother manifold.
Notes:

The spark_vae.ipynb script depends on spark_vae_jacobian.ipynb to calculate the jacobian.
